### PR TITLE
Add notifications for leases

### PR DESCRIPTION
**Pull request by DanNiESh**
In progress

### CODE REVIEW

**Comment by tzumainn** [comment]
One question: is it possible for the Lease payload to include the Ironic Node payload? At the minimum, we probably want to have the node name and uuid and maybe the power state and provisioning state.

**Comment by DanNiESh** [comment]
> he Ironic Node payload?



> One question: is it possible for the Lease payload to include the Ironic Node payload? At the minimum, we probably want to have the node name and uuid and maybe the power state and provisioning state.

Yes, I can add them into lease payload.

File: esi_leap/objects/lease.py [hunk 1]
```diff
@@ -132,14 +202,29 @@ def cancel(self):
                         parent_lease = Lease.get(self.parent_lease_uuid)
                         resource.set_lease(parent_lease)
 
-                self.status = statuses.DELETED
-                self.expire_time = datetime.datetime.now()
+                resource._node_name = resource.get_name()
+                resource._provision_state \
+                    = resource._get_node_attr('provision_state')
+                resource._power_state = resource._get_node_attr('power_state')
```
Comment (tzumainn) [inline esi_leap/objects/lease.py:208]: I don't think we should be calling a `_` method; maybe add `get_*_state` methods?

**Review by tzumainn — CHANGES_REQUESTED** [review]
Looks good! A few comments.

File: esi_leap/objects/lease.py [hunk 2]
```diff
@@ -29,6 +31,74 @@
...
+        'fulfill_time': fields.DateTimeField(nullable=True),
+        'expire_time': fields.DateTimeField(nullable=True),
+        'status': fields.StringField(),
+        'properties': fields.FlexibleDictField(nullable=True),
+        'offer_uuid': fields.UUIDField(nullable=True),
+        'parent_lease_uuid': fields.UUIDField(nullable=True),
+        'node_name': fields.StringField(nullable=True),
+        'node_uuid': fields.UUIDField(),
+        'node_provision_state': fields.StringField(),
+        'node_power_state': fields.StringField(),
```
[the remark this answers is not in the document]
Comment (DanNiESh) [reply esi_leap/objects/lease.py:89]: will do

File: esi_leap/common/rpc.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+from osprofiler import profiler
+
+NOTIFICATION_TRANSPORT = None
+VERSIONED_NOTIFIER = None
+
+ALLOWED_EXMODS = [
+    exception.__name__,
+]
+EXTRA_EXMODS = []
+TOPICS = ['esi_leap_versioned_notifications']
```
[the remark this answers is not in the document]
Comment (DanNiESh) [reply esi_leap/common/rpc.py:26]: Yeah, it could be configurable by adding a new parameter in conf.notification

**Comment by DanNiESh** [comment]
I made changes to the code according to the comments. I will test it on demo server sometime tomorrow.

File: esi_leap/common/rpc.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import oslo_messaging as messaging
+from osprofiler import profiler
+
+NOTIFICATION_TRANSPORT = None
+VERSIONED_NOTIFIER = None
+
+ALLOWED_EXMODS = [
+    exception.__name__,
+]
+EXTRA_EXMODS = []
```
Comment (tzumainn) [inline esi_leap/common/rpc.py:26]: Is this needed?

File: esi_leap/common/rpc.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def init(conf):
+    global NOTIFICATION_TRANSPORT
+    global VERSIONED_NOTIFIER
+    exmods = get_allowed_exmods()
```
Comment (tzumainn) [inline esi_leap/common/rpc.py:32]: can this just be set to ALLOWED_EXMODS?

File: esi_leap/resource_objects/test_node.py [hunk 6]
```diff
@@ -42,8 +42,17 @@ def get_lease_uuid(self):
...
+        return 'available'
+
     def set_lease(self, lease):
         return
 
     def remove_lease(self, lease):
         return
+
+    def _get_node_attr(self, attr):
+        return None
```
Comment (tzumainn) [inline esi_leap/resource_objects/test_node.py:58]: Is this still needed?

File: esi_leap/resource_objects/test_node.py [hunk 7]
```diff
@@ -42,8 +42,17 @@ def get_lease_uuid(self):
     def get_lessee_project_id(self):
         return self._project_id
 
+    def get_node_power_state(self):
+        return 'Off'
+
+    def get_node_provision_state(self):
+        return 'available'
+
```
Comment (tzumainn) [inline esi_leap/resource_objects/test_node.py:50]: Should these methods be added to dummy nodes as well?

File: esi_leap/resource_objects/ironic_node.py [hunk 8]
```diff
@@ -45,6 +45,10 @@ def __init__(self, ident):
         else:
             self._node = None
             self._uuid = ident
+        self._node_name = None
+        self._provision_state = None
+        self._power_state = None
+        self._node_properties = None
```
Comment (tzumainn) [inline esi_leap/resource_objects/ironic_node.py:51]: Ah - notifications won't work on dummy/fake nodes unless they have these values as well, correct? I'm not sure we want to set these values however; adding a comment above.

File: esi_leap/objects/lease.py [hunk 9]
```diff
@@ -211,3 +309,11 @@ def resource_object(self):
...
         return self.dbapi.lease_verify_child_availability(
             self, start_time, end_time)
+
+    def _get_resource_info(self, resource):
+        resource._node_name = resource.get_name()
+        resource._provision_state \
+            = resource.get_node_provision_state()
+        resource._power_state = resource.get_node_power_state()
+        resource._node_properties = resource.get_config()
+        return resource
```
Comment (tzumainn) [inline esi_leap/objects/lease.py:319]: How are these values pulled out by the notification? Setting and using `_` variables is probably not idea, when methods like `resource.get_name()` retrieve the value (and exists for all three node types).

**Review by tzumainn — CHANGES_REQUESTED** [review]
Hi! Some rather scattered comments, plus a bigger concern about the setting of resource information.

File: esi_leap/resource_objects/ironic_node.py (hunk 8 above)
Comment (DanNiESh) [reply esi_leap/resource_objects/ironic_node.py:51]: Yes. Notification only works on ironic nodes currently.

File: esi_leap/objects/lease.py (hunk 9 above)
Comment (DanNiESh) [reply esi_leap/objects/lease.py:319]: The ironic_node and lease objects are passed to populate_schema() https://github.com/DanNiESh/esi-leap/blob/events/esi_leap/objects/notification.py#L132. Then it uses getattr() method to get the node and lease objects' attribute ( _power_state, _provision_state, _uuid, etc.) based on the schema.

File: esi_leap/common/rpc.py (hunk 5 above)
Comment (DanNiESh) [reply esi_leap/common/rpc.py:32]: Sure. I'll update it

File: esi_leap/objects/lease.py (hunk 9 above)
Comment (tzumainn) [reply esi_leap/objects/lease.py:319]: In that case, I might suggest creating something like a new 'resource_event' object that's specifically meant to be passed in, and have this function just instantiate that object.

Comment (DanNiESh) [reply esi_leap/objects/lease.py:319]: Will do

File: esi_leap/resource_objects/test_node.py (hunk 6 above)
Comment (DanNiESh) [reply esi_leap/resource_objects/test_node.py]: I removed it since it's not called anywhere.

File: esi_leap/resource_objects/test_node.py (hunk 7 above)
Comment (DanNiESh) [reply esi_leap/resource_objects/test_node.py:50]: Added. Also added unit test in test_dummy_nodes.py

**Comment by DanNiESh** [comment]
I created a new class LeaseResourceEvent to pass in notification emit method. The new changes has been tested in demo server. A message json body looks like this: 
```
{'message_id': 'f17a25ff-466e-45f7-9e59-5549ab3aa925', 'publisher_id': 'esi-leap-manager.esi-demo-ironic-controller.localdomain', 'event_type': 'esi_leap.LeaseResourceEvent.fulfill.end', 'priority': 'INFO', 'payload': {'esi_leap_object.name': 'LeaseCRUDPayload', 'esi_leap_object.namespace': 'esi_leap', 'esi_leap_object.version': '1.0', 'esi_leap_object.data': {'id': 113, 'name': None, 'uuid': '56a63bcd-43ac-440c-812d-e0f11166fc61', 'project_id': '218b3996b35b4e378f7701c78708d4d7', 'owner_id': '218b3996b35b4e378f7701c78708d4d7', 'resource_type': 'ironic_node', 'resource_uuid': 'f8b86846-86f2-4b63-95bc-fd3dfea93751', 'start_time': '2023-06-03T00:12:53Z', 'end_time': '9999-12-31T23:59:59Z', 'fulfill_time': None, 'expire_time': None, 'status': 'created', 'properties': {}, 'offer_uuid': None, 'parent_lease_uuid': None, 'node_name': 'dell-14', 'node_uuid': 'f8b86846-86f2-4b63-95bc-fd3dfea93751', 'node_provision_state': 'available', 'node_power_state': 'power off', 'node_properties': {'vendor': 'dell inc', 'local_gb': '557', 'cpus': '32', 'cpu_arch': 'x86_64', 'memory_mb': '393216', 'capabilities': 'cpu_vt:true,cpu_aes:true,cpu_hugepages:true,cpu_hugepages_1g:true,cpu_txt:true'}}}, 'timestamp': '2023-06-03 00:13:38.598030', '_unique_id': '1bdb824e578e4a1eb6c9e39bd6e424d7', '_context_user': None, '_context_project_id': None, '_context_system_scope': None, '_context_project': None, '_context_domain': None, '_context_user_domain': None, '_context_project_domain': None, '_context_is_admin': False, '_context_read_only': False, '_context_show_deleted': False, '_context_auth_token': None, '_context_request_id': 'req-f9524840-9c06-42cd-bb43-f434d2ba067d', '_context_global_request_id': None, '_context_resource_uuid': None, '_context_roles': [], '_context_user_identity': '- - - - -', '_context_is_admin_project': True}
```

**Review by tzumainn — CHANGES_REQUESTED** [review]

**Comment by DanNiESh** [comment]
Update: tested on demo server. The notification message is 
```
{'message_id': '74f335a7-221b-46e1-b8f5-08b65dd548fa', 'publisher_id': 'esi-leap-manager.esi-demo-ironic-controller.localdomain', 'event_type': 'esi_leap.lease.fulfill.end', 'priority': 'INFO', 'payload': {'esi_leap_object.name': 'LeaseCRUDPayload', 'esi_leap_object.namespace': 'esi_leap', 'esi_leap_object.version': '1.0', 'esi_leap_object.data': {'id': 114, 'name': None, 'uuid': '412b071a-b37b-4a00-a0b4-ed90643b2d51', 'project_id': '218b3996b35b4e378f7701c78708d4d7', 'owner_id': '218b3996b35b4e378f7701c78708d4d7', 'resource_type': 'ironic_node', 'resource_uuid': 'f8b86846-86f2-4b63-95bc-fd3dfea93751', 'start_time': '2023-06-08T16:42:05Z', 'end_time': '9999-12-31T23:59:59Z', 'fulfill_time': '2023-06-08T16:43:02Z', 'expire_time': None, 'status': 'active', 'properties': {}, 'offer_uuid': None, 'parent_lease_uuid': None, 'node_name': 'dell-14', 'node_uuid': 'f8b86846-86f2-4b63-95bc-fd3dfea93751', 'node_provision_state': 'available', 'node_power_state': 'power on', 'node_properties': {'vendor': 'dell inc', 'local_gb': '557', 'cpus': '32', 'cpu_arch': 'x86_64', 'memory_mb': '393216', 'capabilities': 'cpu_vt:true,cpu_aes:true,cpu_hugepages:true,cpu_hugepages_1g:true,cpu_txt:true'}}}, 'timestamp': '2023-06-08 16:43:02.094990', '_unique_id': '3c3b9b5b725a4b99b7b27147d472cd10', '_context_user': None, '_context_project_id': None, '_context_system_scope': None, '_context_project': None, '_context_domain': None, '_context_user_domain': None, '_context_project_domain': None, '_context_is_admin': False, '_context_read_only': False, '_context_show_deleted': False, '_context_auth_token': None, '_context_request_id': 'req-f90b9c92-5955-47a0-a365-87fe33532457', '_context_global_request_id': None, '_context_resource_uuid': None, '_context_roles': [], '_context_user_identity': '- - - - -', '_context_is_admin_project': True}
```

**Review by tzumainn — APPROVED** [review]

**Comment by tzumainn** [comment]
Looks good - thanks!

**Pull request merged**